sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,634 @@
+ * {@link PCollection PCollection&lt;SolrInputDocument&gt;} (which can be bounded or unbounded).
+ *
+ * <p>To configure {@link SolrIO#write SolrIO.write()}, similar to the read, you
+ * have to provide a connection configuration. For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...)
+ *    .apply(SolrIO.write().withConnectionConfiguration(
+ *       SolrIO.ConnectionConfiguration.create("127.0.0.1:9983", "my-collection")
+ *    )
+ *
+ * }</pre>
+ *
+ * <p>Optionally, you can provide {@code withBatchSize()}
+ * to specify the size of the write batch in number of documents.
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class SolrIO {
+
+  public static Read read() {
+    return new AutoValue_SolrIO_Read.Builder().setBatchSize(100L).setQuery("*:*").build();
+  }
+
+  public static Write write() {
+    return new AutoValue_SolrIO_Write.Builder().setMaxBatchSize(1000L).build();
+  }
+
+  private SolrIO() {
+  }
+
+  /** A POJO describing a connection configuration to Solr. */
+  @AutoValue
+  public abstract static class ConnectionConfiguration implements Serializable {
+
+    abstract String getZkHost();
+    abstract String getCollection();
+    @Nullable abstract String getUsername();
+    @Nullable abstract String getPassword();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setZkHost(String zkHost);
+      abstract Builder setCollection(String collection);
+      abstract Builder setUsername(String username);
+      abstract Builder setPassword(String password);
+      abstract ConnectionConfiguration build();
+    }
+
+    /**
+     * Creates a new Solr connection configuration.
+     *
+     * @param zkHost host of zookeeper
+     * @param collection the collection toward which the requests will be issued
+     * @return the connection configuration object
+     * @throws IOException when it fails to connect to Solr
+     */
+    public static ConnectionConfiguration create(String zkHost, String collection)
+        throws IOException {
+      checkArgument(zkHost != null,
+          "ConnectionConfiguration.create(zkHost, collection) "
+              + "called with null address");
+      checkArgument(collection != null,
+          "ConnectionConfiguration.create(zkHost, collection) "
+              + "called with null collectioin");
+      return new AutoValue_SolrIO_ConnectionConfiguration.Builder()
+          .setZkHost(zkHost)
+          .setCollection(collection)
+          .build();
+    }
+
+    /**
+     * If Solr basic authentication is enabled, provide the username and password.
+     *
+     * @param username the username used to authenticate to Solr
+     * @param password the password used to authenticate to Solr
+     * @return the {@link ConnectionConfiguration} object with basic credentials set
+     */
+    public ConnectionConfiguration withBasicCredentials(String username, String password) {
+      checkArgument(
```

Comment:
Ugh I really don't like this style. But ok to keep as-is for this PR, I'll raise it on dev@.